### ISSUE TITLE
Exits badly (stack trace) on some external links

### ISSUE THREAD

**Issue by danielfdickinson** [Munter/hyperlink#196]
When I don't use the ``-i`` (internal links only) on a recursive (``-r``) check of the root ``index.html`` on one of the pages on my site, I get:

```
ok 1430 load https://docs.ansible.com/ansible/latest/user_guide/vault.html
/workspaces/hugo-sites/hugo-www-wildtechgarden-ca/node_modules/hyperlink/lib/index.js:535
          asset.parseTree.querySelectorAll('[id]')
                          ^

TypeError: Cannot read properties of undefined (reading 'querySelectorAll')
    at processAsset (/workspaces/hugo-sites/hugo-www-wildtechgarden-ca/node_modules/hyperlink/lib/index.js:535:27)
    at runNextTicks (node:internal/process/task_queues:61:5)
    at processImmediate (node:internal/timers:437:9)
```
The offending link was: ``https://yaml.org/spec/1.1/#escaping%20in%20double-quoted%20style/`` as that id no longer exists in the document (they've replaced them with #idxxxxxx style ids), but the document still does).

The same message occurs when a page with a fragment has gone 404: e.g. ``https://docs.oracle.com/en/cloud/paas/database-dbaas-cloud/csdbi/generate-ssh-key-pair.html#GUID-4285B8CF-A228-4B89-9552-FE6446B5A673``.

**Comment by papandreou** [comment]
Tried applying a fix based on the stack trace, could you try 5.0.1 and LMK if that's better?

**Comment by danielfdickinson** [comment]
Thanks for the quick push! The first link no longer has an issue, but second link now gives:
```
ok 1430 unloading public/images/xca-new-certificate-dialogue-with-tls-server-template-selected.png
addAsset [undefined/74212 https://docs.oracle.com/pls/rdman/handle404?X_ORCL_404path=%2fen%2fcloud%2fpaas%2fdatabase-dbaas-cloud%2fcsdbi%2fgenerate-ssh-key-pair.html]
ok 1431 load https://docs.oracle.com/en/cloud/paas/database-dbaas-cloud/csdbi/generate-ssh-key-pair.html
/workspaces/hugo-sites/hugo-www-wildtechgarden-ca/node_modules/hyperlink/lib/index.js:538
            asset.parseTree.querySelectorAll('[id]')
                            ^

TypeError: Cannot read properties of undefined (reading 'querySelectorAll')
    at processAsset (/workspaces/hugo-sites/hugo-www-wildtechgarden-ca/node_modules/hyperlink/lib/index.js:538:29)
    at runNextTicks (node:internal/process/task_queues:61:5)
    at processImmediate (node:internal/timers:437:9)
```

**Comment by papandreou** [comment]
Urgh, try with 5.0.2.

(If that fixes it, I'll try to make a proper test case 😇 )

**Comment by danielfdickinson** [comment]
:tada: That did it! Thank you again.

**Comment by danielfdickinson** [comment]
There is a gotcha now. I'm not sure if it's new or just previously hidden but an external link with a fragment that does exist is treated as an internal link and hyperlink crawls that external site as if it were internal (despite ``--canonicalroot``)

Would you like a separate issue for that?

**Comment by papandreou** [comment]
Hmm, could you share a complete case that reproduces that? I tried building one, but it didn't happen.

**Comment by danielfdickinson** [comment]
Absolutely. I'll create a new issue for it too. It might be a day or two by the time we sync up (I'm hoping to get to it later tonight, but I'm on the other side of the Atlantic so by the time you see it, it will be tomorrow, and that's assuming I can). In the meantime the offending link is: <https://docs.ovh.com/gb/en/dedicated/firewall-network#objective> in case it is a site-specific issue.

**Comment by papandreou** [comment]
Thanks! That helps. Looks like what's triggering the problem is that there's a 301 redirect involved (adding a `/` to the end):

```
$ curl -i https://docs.ovh.com/gb/en/dedicated/firewall-network
HTTP/2 301 
date: Sun, 21 Nov 2021 08:21:26 GMT
content-type: text/html
location: https://docs.ovh.com/gb/en/dedicated/firewall-network/
```

I'll take a look, no need for a separate issue :+1:

**Comment by papandreou** [comment]
Hopefully fixed in 5.0.3. Thanks for the reports and extra details! :bow:

**Comment by danielfdickinson** [comment]
Works perfectly now! Thank you. :bow: :tada:
